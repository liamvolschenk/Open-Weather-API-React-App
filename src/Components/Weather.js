import React, { useState } from 'react';

//setting the api as a variable so its easier to use later
const api = {
  key: "15c4df57e4b67b30dd3dfeb1be4e0aed",
  base: "https://api.openweathermap.org/data/2.5/"
}

function Weather() {

  //setting the state for the query and weather
  const [query, setQuery] = useState('');
  const [weather, setWeather] = useState({});

  //here I fetch the API when enter is pressed and return it as a json promise, and use the result to set the state
  const search = evt => {
    if (evt.key === "Enter") {
      fetch(`${api.base}weather?q=${query}&units=metric&APPID=${api.key}`)
        .then(res => res.json())
        .then(result => {
          setWeather(result);
          setQuery('');
        });
    }
  }

  //This is a basic function to fetch the date and output it on the users screen when they search 
  const dateBuilder = (d) => {
    let months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
    let days = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

    let day = days[d.getDay()];
    let date = d.getDate();
    let month = months[d.getMonth()];
    let year = d.getFullYear();

    return `${day} ${date} ${month} ${year}`
  }

  return (
    <div data-testid="test-1" className={(typeof weather.main != "undefined") ? ((weather.main.temp > 19) ? 'app warm' : 'app') : 'app'}>    {/*this sets the background image based on the temp, if it is warmer than 19°C it will change to the 'warm background'*/}
      <main>
        <div className="heading">
          Search below for a town or city to see their current weather:
        </div>
        <div className="search-box">
          <input
            type="text"
            className="search-bar"
            placeholder=" Search..."
            onChange={e => setQuery(e.target.value)}
            value={query}
            onKeyPress={search}
          />
        </div>
        {(typeof weather.main != "undefined") ? (
          <div>
            <div className="location-box">
              <div className="location">{weather.name}, {weather.sys.country}</div>
              <div className="date">{dateBuilder(new Date())}</div>
            </div>
            <div className="weather-box">
              <div className="temp">
                {Math.round(weather.main.temp)}°C
              </div>
              <div className="weather">
                {weather.weather[0].main}
              </div>
            </div>
          </div>
        ) : ('')}
      </main>
    </div>
  );
}

export default Weather;
