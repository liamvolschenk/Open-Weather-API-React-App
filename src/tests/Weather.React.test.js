import React from 'react';
import renderer from 'react-test-renderer';

import Weather from '../Components/Weather';

//Snapshot Test
test("snapshot for weather component", () => {
    const tree = renderer.create(<Weather />).toJSON();
    expect(tree).toMatchSnapshot();
})

//API fetch function test
async function withFetch() {
    const res = await fetch("https://api.openweathermap.org/data/2.5/")
    const json = await res.json()
  
    return json
  }
  
  const unmockedFetch = global.fetch
  
  beforeAll(() => {
    global.fetch = () =>
      Promise.resolve({
        json: () => Promise.resolve([]),
      })
  })
  
  afterAll(() => {
    global.fetch = unmockedFetch
  })
  
  
  describe('Weather', () => {
    test('fetch test', async () => {
      const json = await withFetch()
      
      expect.extend({
        yourMatcher() {
          return {
            pass: true,
            message: () => 'error',
          };
        },
      });
    })
  })