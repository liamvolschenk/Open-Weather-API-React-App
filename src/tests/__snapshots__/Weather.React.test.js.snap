// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for weather component 1`] = `
<div
  className="app"
  data-testid="test-1"
>
      
  <main>
    <div
      className="search-box"
    >
      <input
        className="search-bar"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Search..."
        type="text"
        value=""
      />
    </div>
    
  </main>
</div>
`;
